[SVCS-599] Fix OAuth Client Revoke Failures

cslzchen opened the pull request:
### Ticket

[SVCS-599](https://openscience.atlassian.net/browse/SVCS-599)

### Purpose, Changes, Side Effects and More

In short, the issue is fixed by deprecating `revokeClientTokens()` and move its functionality to the controller `OAuth20RevokeClientTokensController`. Please refer to the ticket for detailed information and discussion.

Note: This PR replaces https://github.com/CenterForOpenScience/cas-overlay/pull/91.

cslzchen committed:
Fix oauth revoke - part 1

- Remove @Transient from getTicket for accessToken
- Explicilty declare JPA access type as FIELD

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/token/AccessTokenImpl.java:
```diff
@@ -20,29 +20,34 @@
  */
 @Entity
 @Table(name="ACCESSTOKEN")
+@Access(AccessType.FIELD)
```
Explicitly set the default access type.

cslzchen replied:
Explicitly set the default access type.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/token/AccessTokenImpl.java:
```diff
@@ -93,12 +105,10 @@ public AccessTokenImpl(final String id, final TokenType type, final String clien
     }
 
     @Override
-    @Transient
```
~~This is what was causing the error. Although this looks like a getter but there is no field for `ticket` at all. Not sure how `@Transient` apply to an non-getter.~~ Update: this only causes 1/3 of the problems. All fixed with the new solution.

cslzchen reviewed (commented):
Work In Progress. Need a few more tests before ready for CR.

icereval commented:
:+1:

cslzchen reviewed (commented):
~~The code is tested locally with full OAuth flow working as expected. Ready for review @icereval .~~

- Revoking clients with ~~multiple access tokens and refresh tokens~~ one refresh token and multiple access tokens obtained from that refresh token succeeds.
- Without `@Transient`, `.getTicket()` acts like a helper/utility method for all related classes and doesn't do anything else. This is the expected behavior.
- [x] Not sure why we had `@Transient` in the first place (commit message does not help). Please double check before approving/merging.

icereval reviewed (approved):
LGTM

cslzchen commented:
Update: further tests revealed a few more issues. Please refer to the ticket for more detail.

icereval reviewed (changes requested):
Holding on merge for further testing.

cslzchen committed:
Fix oauth revoke - part 2

The issue is fixed by only revoke one token at a time in the
callee: `CentralOAuthServiceImpl.revokeClientTokens()` and
move the `for` loop and the service check to the caller:
`OAuth20RevokeClientTokensController.handleRequestInternal()`.
`revokeClientTokens()` returns `Boolean.TRUE` if all tokens
are cleared or `Boolean.FASLE` if there is more to continue.

cslzchen committed:
Update tests for revoke client tokens

cslzchen committed:
Improve token revoking

- Removed revokeClientTokens() from the interface,
  the classe and the controller
- Exposed tokenRegistry from CentralOAuthService
  to OAuth20RevokeClientTokensController
- Removed unecessary while loop and the iteration
  threshold
- The controller now uses a better for loop which
  iteration the token list and remove each of them

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/CentralOAuthService.java:
```diff
@@ -216,4 +222,9 @@ AuthorizationCode grantAuthorizationCode(TokenType type, String clientId, String
+    /**
+     * @return the token registry.
+     */
+    TokenRegistry getTokenRegistry();
```
Expose the `tokenRegistry` in `CentralOAuthService` to `OAuth20RevokeClientTokensController`.

cslzchen replied:
Update: a better interface design should hide the token registry but return a list of available tokens associated with the client.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/CentralOAuthService.java:
```diff
@@ -123,14 +125,18 @@ AuthorizationCode grantAuthorizationCode(TokenType type, String clientId, String
+//     * @param clientSecret the client secret
+//     * @return a Boolean status if tokens were successfully revoked.
+//     */
+//    Boolean revokeClientTokens(String clientId, String clientSecret);
```
Deprecating the `revokeClientTokens()` which is no longer used. For future references, I decide to keep it in the code base commented out for now and for a while.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/CentralOAuthServiceImpl.java:
```diff
@@ -286,32 +291,33 @@ public Boolean revokeToken(final Token token) {
-    }
+//    TODO: Deprecated. Please remove.
+//    @Override
+//    public Boolean revokeClientTokens(final String clientId, final String clientSecret) {
```
Similarly to the interface, the implementation of `revokeClientTokens()` is deprecated.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20RevokeClientTokensController.java:
```diff
@@ -68,11 +74,29 @@ protected ModelAndView handleRequestInternal(final HttpServletRequest request, f
+        }
+
+        // Retrieve the tokens from the registry and remove all of them
+        TokenRegistry tokenRegistry = centralOAuthService.getTokenRegistry();
```
Use the exposed `tokenRegistry` to get a list of all tokens. Remove them one by one. No more bad `while` loops.

cslzchen reviewed (commented):
Move to RTM after another round of full test and before/after verification.

cslzchen committed:
Expose token list instead of token registry

- The oauth controller only needs a list of tokens
  associated with the specified client id.  It is
  overkill to provide it with the token registry.
  In addition, from the point of lease previlige,
  the controller should only be provided what it
  needs to finish the job.  No more no less.
- Added `getClientAccessTokens()` and
  `getClientRefreshTokens()` to the interface and
  implements them.
- Removed `getTokenRegistry()` and its implementation
  from the interface and the class respectively.
- Updated the controller.

cslzchen committed:
Deprecate `getClientTokens()` in a better way

- Keep it in the interface with extra docstring
- Change the return type from `Boolean` to `void`
- Throws `UnsupportedOperationException` in the
  implementation with a useful error message

cslzchen committed:
Updated controller tests and update docstrings

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/CentralOAuthService.java:
```diff
@@ -126,11 +127,13 @@ AuthorizationCode grantAuthorizationCode(TokenType type, String clientId, String
-     * @return a Boolean status if tokens were successfully revoked.
      */
-    Boolean revokeClientTokens(String clientId, String clientSecret);
+     void revokeClientTokens(String clientId, String clientSecret);
```
Deprecate `revokeClientTokens()`.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/CentralOAuthServiceImpl.java:
```diff
@@ -287,30 +293,8 @@ public Boolean revokeToken(final Token token) {
-
-        return Boolean.TRUE;
+    public void revokeClientTokens(final String clientId, final String clientSecret) {
+        throw new UnsupportedOperationException("revokeClientTokens are deprecated and should not be used");
```
Deprecate the implementation of `revokeClientToken()`.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/token/RefreshTokenImpl.java:
```diff
@@ -39,6 +41,7 @@
  */
 @Entity
 @Table(name="REFRESHTOKEN")
+@Access(AccessType.FIELD)
```
Explicitly set the default access type for both refresh and access token.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20RevokeClientTokensController.java:
```diff
@@ -68,11 +73,29 @@ protected ModelAndView handleRequestInternal(final HttpServletRequest request, f
             return OAuthUtils.writeJsonError(response, OAuthConstants.INVALID_REQUEST, e.getMessage(), HttpStatus.SC_BAD_REQUEST);
         }
 
-        if (!centralOAuthService.revokeClientTokens(clientId, clientSecret)) {
```
The functionality of `revokeClientTokens()` are removed and directly handled by the controller here in `OAuth20RevokeClientTokensController.handleRequestInternal()`. Please refer to the code below and the following four comments

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20RevokeClientTokensController.java:
```diff
@@ -68,11 +73,29 @@ protected ModelAndView handleRequestInternal(final HttpServletRequest request, f
-        if (!centralOAuthService.revokeClientTokens(clientId, clientSecret)) {
+        // Verify that the client service has a valid client id and client secret
+        final OAuthRegisteredService service = centralOAuthService.getRegisteredService(clientId);
+        if (service == null || !service.getClientSecret().equals(clientSecret)) {
```
Validate service first.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20RevokeClientTokensController.java:
```diff
@@ -68,11 +73,29 @@ protected ModelAndView handleRequestInternal(final HttpServletRequest request, f
+        }
+
+        // Remove all refresh tokens for the client of the specified id
+        final Collection<RefreshToken> refreshTokens = centralOAuthService.getClientRefreshTokens(clientId);
```
Remove all refresh tokens.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/CentralOAuthService.java:
```diff
@@ -216,4 +219,20 @@ AuthorizationCode grantAuthorizationCode(TokenType type, String clientId, String
+     * @param clientId the client id
+     * @return a list of refresh tokens
+     */
+    Collection<RefreshToken> getClientRefreshTokens(String clientId);
```
`CentralOauthService` now exposes the refresh token list of a specified client to `OAuth20RevokeClientTokensController`. Same for the access token list below.

cslzchen commented on cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20RevokeClientTokensController.java:
```diff
@@ -68,11 +73,29 @@ protected ModelAndView handleRequestInternal(final HttpServletRequest request, f
+            centralOAuthService.revokeToken(token);
+        }
+        // Remove all access tokens for the client of the specified id
+        final Collection<AccessToken> accessTokens = centralOAuthService.getClientAccessTokens(clientId);
```
Remove all access tokens.

cslzchen reviewed (commented):
PR ready again 🔥 🔥 

- Tested locally with three different scenarios. All passed Postman tests and further verified inside the database.
- Before/After test for DB passes as well. (There is no actual model change.)

Please refer to the ticket for more detailed test configuration and results.

cslzchen reviewed (commented):
cc @icereval , I did another round of local test just now and all look good. Merged 🎆!
